File: src/characters/fortune_teller.ml
open! Core
open! Import

include Character.Make (struct
    type t =
      { kind : Kind.t
      ; alignment : Alignment.t
      }

    let id = "fortune_teller"
    let name = "Fortune Teller"
    let t = { kind = Kind.Townsfolk; alignment = Alignment.Good }
    let kind { kind; _ } = kind
    let alignment { alignment; _ } = alignment
  end)

let night_action ~player_id:pid ~night:_ =
  Some
    (if_alive
       pid
       (let%bind.Botc_exec state = Botc_exec.get_state
        and () = Botc_exec.wake pid in
        let candidates = Game_state.alive_ids state in
        let%bind.Botc_exec p1 = Botc_exec.ask pid "Choose a player" candidates in
        let%bind.Botc_exec p2 = Botc_exec.ask pid "Choose a player" candidates in
        let result =
          if Game_state.is_poisoned state pid
          then false
          else
            Kind.equal (Game_state.kind state p1) Kind.Demon
            || Kind.equal (Game_state.kind state p2) Kind.Demon
        in
        let%bind.Botc_exec () = Botc_exec.tell pid (if result then "Yes" else "No") in
        Botc_exec.sleep pid))
;;

let day_action ~player_id:_ = None
let on_setup ~player_id:_ = None
let on_nominated ~player_id:_ ~nominator:_ = None
let on_executed ~player_id:_ = None
let on_night_kill ~player_id:_ = None

File: src/characters/fortune_teller.rs
//! Fortune Teller (Townsfolk): each night, choose two players and learn if
//! either registers as the Demon. One good player is a "red herring" who always
//! registers as the Demon to you.

use crate::ability::Ability;
use crate::ctx::Ctx;
use crate::ids::PlayerId;
use crate::role::{Alignment, CharacterInfo, Kind};
use crate::token::Token;

character!(
    FortuneTeller,
    "fortune_teller",
    "Fortune Teller",
    Kind::Townsfolk,
    Alignment::Good
);

impl Ability for FortuneTeller {
    fn info(&self) -> CharacterInfo {
        Self::INFO
    }

    fn on_setup(&self, ctx: &mut Ctx, me: PlayerId) {
        // The storyteller marks one good player as the red herring.
        let good: Vec<PlayerId> = ctx
            .grim
            .seats()
            .filter(|&id| !ctx.grim.get(id).alignment.is_evil())
            .collect();
        if good.is_empty() {
            return;
        }
        let labels: Vec<String> = good.iter().map(|&id| ctx.label(id)).collect();
        let i = ctx.choose(&format!("red herring for {}", ctx.label(me)), &labels);
        ctx.grim.add_token(good[i], Token::RedHerring);
    }

    fn on_night(&self, ctx: &mut Ctx, me: PlayerId, _night: u32) {
        if !ctx.grim.is_alive(me) {
            return;
        }
        ctx.wake(me);
        let candidates: Vec<PlayerId> = ctx.grim.alive().collect();
        let p1 = ctx.ask(me, "Choose a player", &candidates);
        let p2 = ctx.ask(me, "Choose a player", &candidates);
        let truth = ctx.registers_as_demon(p1, "fortune_teller")
            || ctx.registers_as_demon(p2, "fortune_teller");
        let shown = ctx.deliver(me, yes_no(truth), &["Yes", "No"]);
        ctx.reveal(me, shown);
        ctx.sleep(me);
    }
}

fn yes_no(b: bool) -> &'static str {
    if b {
        "Yes"
    } else {
        "No"
    }
}
